### ISSUE TITLE
Add 'async' methods where applicable

### ISSUE THREAD

**Issue by ianschmitz** [recurly/recurly-client-net#119]
Working with ASP.NET Web API, we follow the .NET async await pattern for any I/O bound work. This greatly increases the load capacity/scale of our web tier.

I would love to see the Recurly .NET client expose asynchronous methods that return `Task` or `Task<T>`. This will allow me to `await` the returned task and free up the thread to serve other requests.

**Comment by bhelx** [comment]
I can definitely see wanting this feature. I'm wondering if there would be a simple way to implement it without changing the whole API or providing some kind of second, async api. Any ideas?

**Comment by MuffinMan10110** [comment]
This would definitely be a convenient feature, however there is one big issue preventing this from happening (and some concerns):
- The library currently targets .NET 3.5 and the System.Threading.Tasks API did not come out until .NET 4. 
- Upgrading the library from 3.5 to 4.0 breaks existing users who aren't targeting 4.0 and limits those in the future that choose to use Recurly on an older app that targets <= 3.5.
- Minor, but we probably need to write and support additional tests for all async methods (even though they would just be wrappers around the existing synchronous code)

Your best bet is to wrap the create account call in an async method, something along the lines of:
     
    Task<Account>CreateAccountAsync(Account account)
    {
        return Task.Run(() => account.Create());
    }

That example might be .NET 4.5, but the concept still applies with the 4.0 API.

**Comment by ianschmitz** [comment]
The general pattern I've experienced with .NET and 3rd party libraries, is they generally will expose both synchronous and asynchronous methods for a given operation. 

For example using Recurly, right now you have:
`public static Account Get(string accountCode);`

Following the async pattern i've experienced you would expose both:
```
public static Account Get(string accountCode);
public static Task<Account> Get(string accountCode);
```

**Comment by ianschmitz** [comment]
Sorry i didn't see your post @MuffinMan10110 before i submitted. You have raised some great points!

This isn't my area of expertise, but i believe you can add the TAP (task-based asynchronous pattern) pattern to .NET 3.5 by using `Task Parallel Library for .NET 3.5` as a NuGet package (https://www.nuget.org/packages/TaskParallelLibrary/)

**Comment by MuffinMan10110** [comment]
@ianschmitz Yes I've seen that package before.

I think if you'd like to contribute in adding this feature it would be a welcomed addition.

**Comment by bhelx** [comment]
@ianschmitz I'm fine with it was well as long as it's relatively straightforward to maintain both interfaces.

**Comment by MuffinMan10110** [comment]
Since the library had to be updated to .NET 4.5 for TLS support, we now get all of the latest async api's (Tasks namespace), which would make this effort much easier.

**Comment by BenjaminGibbs** [comment]
The Async/Await pattern increase the capacity/scale of web servers but its a requirement when using frameworks like Orleans or Service Fabric Actors.

My interest is because we are using Orleans so have to use the Async/Await pattern so either I could build my own wrappers around this library for just the bits we need or could look at adding support directly in to this repo so others could use it as well.

Has any progress been made with this issue as of yet?

**Comment by MuffinMan10110** [comment]
@BenjaminGibbs No progress has been made yet. Now that the library is .net 4.5 adding async should be easy. All that we ask is that the library remain backwards compatible. You'd probably need to make an async method that calls the sync method implementation.

**Comment by ianschmitz** [comment]
Be careful going down that road of wrapping synchronous methods with async. You achieve parallelism but not scalability.

There's good reading on this by Stephen Toub who's an async guru: https://blogs.msdn.microsoft.com/pfxteam/2012/03/24/should-i-expose-asynchronous-wrappers-for-synchronous-methods/

**Comment by bhelx** [comment]
@BenjaminGibbs I agree with @MuffinMan10110 there. I'd love to see it as a feature but I have to admit that my priority with this library right now is getting it up to date with the API. It's still a little behind in terms of feature development. After that, hopefully I have some time to work on some of the technical debt. I'm also more than willing to do review of any PRs that come my way. I can always make time to review and test code.

**Comment by BenjaminGibbs** [comment]
I am happy to make a start on this but would be interested to hear what approach you guys think we should be taking? Looking at the codebase `client.cs` is where the network requests are done. So is it a case of having a Sync and Async version of the methods within this, or making (where applicable) everything async under the hood, then having sync wrappers around it to keep backwards compatibility?

If we were going down the route of converting the internals to async, then having wrappers for the API it shouldn't be a too much of a job to make the changes. Here is how I could see it working....

Example taken from `account.cs`

Current:
```c#
    /// <summary>
        /// Delete an account's billing info.
        /// </summary>
        public void DeleteBillingInfo()
        {
            Client.Instance.PerformRequest(Client.HttpRequestMethod.Delete,
                UrlPrefix + Uri.EscapeUriString(AccountCode) + "/billing_info");
            _billingInfo = null;
        }
```
After making it async:
```c#
    /// <summary>
        /// Delete an account's billing info.
        /// </summary>
        public void DeleteBillingInfo()
        {
            var t = Task.Factory.StartNew(async () =>
            {
                await Client.Instance.PerformRequestAsync(Client.HttpRequestMethod.Delete,
                    UrlPrefix + Uri.EscapeUriString(AccountCode) + "/billing_info");
            });
            t.Wait();
            _billingInfo = null;
        }

        /// <summary>
        /// Delete an account's billing info.
        /// </summary>
        async public void DeleteBillingInfoAsync()
        {
            await Client.Instance.PerformRequestAsync(Client.HttpRequestMethod.Delete,
              UrlPrefix + Uri.EscapeUriString(AccountCode) + "/billing_info");
            _billingInfo = null;
        }
```

Then in make Client.cs all async under the hood, some snippets of key changes you would make:

Update the PerformRequest to Async functions.
```c#
 public async Task<HttpStatusCode> PerformRequestAsync(HttpRequestMethod method, string urlPath)
        {
            return await PerformRequestAsync(method, urlPath, null, null, null, null);
        }

    protected virtual async Task<HttpStatusCode> PerformRequestAsync(HttpRequestMethod method, string urlPath,
               WriteXmlDelegate writeXmlDelegate, ReadXmlDelegate readXmlDelegate, ReadXmlListDelegate readXmlListDelegate, ReadResponseDelegate reseponseDelegate)
        {
//---------------------------
}
```

Switch to using `GetRequestStreamAsync` and `GetResponseAsync` 

```c#
 using (var requestStream = await request.GetRequestStreamAsync())
                {
                    WritePostParameters(requestStream, writeXmlDelegate);
                }
//---------------------------

                var r = (HttpWebResponse)await request.GetResponseAsync();
                byte[] pdf;
                var buffer = new byte[2048];
                if (!request.HaveResponse || r.StatusCode != HttpStatusCode.OK) return null;
                using (var ms = new MemoryStream())
                {
                    using (var reader = new BinaryReader(r.GetResponseStream(), Encoding.Default))
                    {
                        int bytesRead;
                        while ((bytesRead = reader.Read(buffer, 0, 2048)) > 0)
                        {
                            ms.Write(buffer, 0, bytesRead);
                        }
                    }
                    pdf = ms.ToArray();
                }
                return pdf;
```

On the whole I think that is approach would keep most of the library 'as-is' in the way it works from an API point of view, allowing in place upgrades for everyone using it currently but would also enable the use of Async throughout while at the same time not have much duplicate code under the hood.

I am open to thoughts on other ways we could do this, or a different preferred approach.

**Comment by BenjaminGibbs** [comment]
Any thoughts @bhelx @MuffinMan10110 ?

**Comment by MuffinMan10110** [comment]
@ianschmitz Thanks for that link on wrapping sync methods with async. It lead me down a few days of research and trying to fully understand what the author is explaining.

@BenjaminGibbs I believe the creation of a new thread via Task.Run is unnecessary here:
```
        public void DeleteBillingInfo()
        {
            var t = Task.Factory.StartNew(async () =>
            {
                await Client.Instance.PerformRequestAsync(Client.HttpRequestMethod.Delete,
                    UrlPrefix + Uri.EscapeUriString(AccountCode) + "/billing_info");
            });
            t.Wait();
            _billingInfo = null;
        }
```
I believe you can just do `PerformRequestAsync(..).Result`.

**Comment by BenjaminGibbs** [comment]
@MuffinMan10110 yes you are correct, I'm too busy living in an async world where normally everything I do normally always triggers a new task - sorry about that.

Another interesting read on tasks and around how the [scheduler works.](http://dotnet.github.io/orleans/Documentation/Advanced-Concepts/External-Tasks-and-Grains.html?q=tasks) - it is in the context of Orleans, but the theory is still the same.

So do we think the above suggestion is a good starting point? or is there another approach you would prefer to see?

I am happy to spend a few hours making the changes and pushing a PR for you guys to look at and see how we could improve on it from there if it is in the right direction.

**Comment by MuffinMan10110** [comment]
@BenjaminGibbs Thanks for sharing that link, it's the first time I've heard of Orleans. It sounds like you've got a good grasp on async so I'm looking forward to see what you come up with.

**Comment by bhelx** [comment]
@BenjaminGibbs It sounds good to me. I do have experience with async architectures, but not in the .NET world. So, I can't really say how it should look. My only requirement is that it doesn't break the API for current customers. As long as we can meet that, I am more than happy to get your PR in.

**Comment by MuffinMan10110** [comment]
@BenjaminGibbs So I did a lot of reading on async/await and what I've learned is that IO and remote requests such as HTTP are great for async because we can give back control to the system while we are waiting for a response. The library doesn't get faster, it just scales available resources further. If the consumer wants to to make the library calls completely not blocking it would be up to it to wrap any calls with `Task.Run`, not the library internally, the reason being that while using Task.Run everywhere is great on a desktop app, it's not going to scale well on a web server.

**Comment by 1TheMuffinMan** [comment]
@ianschmitz I wanted to let you know that we've been busy working on async support. Development on it is finished and is in the testing phase. @bhelx or I will update this issue once we're ready to release it.

**Comment by bhelx** [comment]
Thanks for the update @1TheMuffinMan!

**Comment by ianschmitz** [comment]
Hey thanks for the heads up @1TheMuffinMan, I lost track of this issue.

> So I did a lot of reading on async/await and what I've learned is that IO and remote requests such as HTTP are great for async because we can give back control to the system while we are waiting for a response. The library doesn't get faster, it just scales available resources further.

Precisely! Freeing up a given thread to perform other work while it's busy waiting on an I/O bound task is magical. 

Glad to hear things have progressed on this issue. I'm sure many other .NET devs will be happy.

**Comment by argystu** [comment]
Any update on when this might be released?

**Comment by 1TheMuffinMan** [comment]
@argystu I've started on it and it's progressing slowly. We've been prioritizing changes from the new API versions lately. With that being said there is no ETA, but considering it will take some time to complete, it's possible that we may pre-release with a separate nuget package with a subset of the methods converted to Async variants. I can keep this issue updated as new updates come up.

**Comment by bhelx** [comment]
There hasn't been any progress on this in over a year and we aren't planning on making any major changes to this library. If Async support is something you are interested in, I'd suggest contacting us about the V3 Beta which supports dotnet core and async requests. See [this comment](https://github.com/recurly/recurly-client-net/issues/152#issuecomment-403611308) about getting on the beta.

**Comment by bhelx** [comment]
There hasn't been any progress on this in over a year and we aren't planning on making any major changes to this library. If Async support is something you are interested in, I'd suggest contacting us about the V3 Beta which supports dotnet core and async requests. See [this comment](https://github.com/recurly/recurly-client-net/issues/152#issuecomment-403611308) about getting on the beta.

**Comment by bhelx** [comment]
Everyone: Wanted to drop a follow up in here. The next major version of our API and this library is underway. We've released the new dotnet client as part of our "Early Access" program and it supports dotnet core as well as async calls. If you are interested in trying it out, you can find some information here: https://dev.recurly.com/v2.19/page/recurly-v3-api-early-access
I'd also suggest you email domino@recurly.com and ben@recurly.com to discuss how we can support you.
